pageserver/src/layered_repository/inmemory_layer.rs

```diff
@@ -259,13 +259,17 @@ impl Layer for InMemoryLayer {
     /// Nothing to do here. When you drop the last reference to the layer, it will
     /// be deallocated.
     fn delete(&self) -> Result<()> {
-        Ok(())
+        panic!("can't delete an InMemoryLayer")
     }
 
     fn is_incremental(&self) -> bool {
         self.incremental
     }
 
+    fn is_open(&self) -> bool {
```

Comment:
Let's be careful with the terms. Admittedly we haven't written these down anywhere, but this is how I think about these (I'll go and add these to the glossary):

Open: a layer where new WAL records can be appended to.
Closed: a layer that is read-only, no new WAL records can be appended to it.
Historical: synonym for closed
In-memory: A layer that is kept in memory, and will go away at a crash
On-disk: A layer that is stored on disk. If its end-LSN is older than disk_consistent_lsn, it is known to be fully flushed and fsync'd to local disk. (and S3?)
Frozen layer: an in-memory layer that is Closed.

With this terminology, the problem happens when GC removes a Closed In-memory layer, aka. a Frozen In-Memory layer. So long story short, let's call this function `is_on_disk` or `is_in_memory`. When an in-memory layer is frozen, it's no longer Open.

(With https://github.com/zenithdb/zenith/pull/864, the in-memory layers are also spilled to disk, so I think we'll need to come up with new terms. Ephemeral vs. Permanent, maybe?)